polars/polars-io/src/ndjson_core/ndjson.rs

```diff
@@ -0,0 +1,289 @@
+    ) -> Result<CoreJsonReader<'a>> {
+        let reader_bytes = reader_bytes;
+
+        let schema = match schema {
+            Some(schema) => Cow::Borrowed(schema),
+            None => {
+                let bytes: &[u8] = &reader_bytes;
+                let mut cursor = Cursor::new(bytes);
+
+                let data_type = ndjson::read::infer(&mut cursor, infer_schema_len).unwrap();
+                let schema: polars_core::prelude::Schema =
+                    StructArray::get_fields(&data_type).into();
+
+                Cow::Owned(schema)
+            }
+        };
+        Ok(CoreJsonReader {
+            reader_bytes: Some(reader_bytes),
+            schema,
+            sample_size,
+            n_rows,
+            n_threads,
+            chunk_size,
+            low_memory,
+        })
+    }
+    fn parse_json(&mut self, mut n_threads: usize, bytes: &[u8]) -> Result<DataFrame> {
+        let mut total_rows = 128;
+
+        if let Some((mean, std)) = get_line_stats(bytes, self.sample_size) {
+            total_rows = (bytes.len() as f32 / (mean - 0.01 * std)) as usize;
+        }
+
+        if total_rows == 128 {
+            n_threads = 1;
+        }
+
+        // let n_chunks = total_rows / self.chunk_size;
+
+        let rows_per_thread = total_rows / n_threads;
+
+        let max_proxy = bytes.len() / n_threads / 2;
+        let capacity = if self.low_memory {
+            self.chunk_size
+        } else {
+            std::cmp::min(rows_per_thread, max_proxy)
+        };
+
+        let expected_fields = &self.schema.len();
+        let file_chunks =
+            get_file_chunks(bytes, n_threads, *expected_fields, SEP, Some(QUOTE_CHAR));
+        // If the number of threads given by the user is lower than our global thread pool we create
+        // new one.
+        let owned_pool;
+        let pool = if POOL.current_num_threads() != n_threads {
+            owned_pool = Some(
+                ThreadPoolBuilder::new()
+                    .num_threads(n_threads)
+                    .build()
+                    .unwrap(),
+            );
+            owned_pool.as_ref().unwrap()
+        } else {
+            &POOL
+        };
+
+        let dfs = pool.install(|| {
+            file_chunks
+                .into_par_iter()
+                .map(|(bytes_offset_thread, stop_at_nbytes)| {
+                    let mut read = bytes_offset_thread;
+
+                    let mut last_read = usize::MAX;
+
+                    let mut buffers = init_buffers(&self.schema, capacity)?;
+
+                    loop {
+                        if read >= stop_at_nbytes || read == last_read {
+                            break;
+                        }
+                        let local_bytes = &bytes[read..stop_at_nbytes];
+
+                        last_read = read;
+                        read += parse_lines(local_bytes, &mut buffers)?;
+                    }
+                    let df = DataFrame::new_no_checks(
```

Comment:
Maybe use `DataFrame::new` here. It does a few cheap sensibility checks. It is still zero copy.